README
pdfresurrect
------------
PDFResurrect is a tool aimed at analyzing PDF documents.  The PDF format allows
for previous document changes to be retained in a more recent version of the
document, thereby creating a running history of changes for the document.  This
tool attempts to modify the PDF so that a reading utility will be presented with
the previous versions of the PDF.  The modified "versions" will be generated
as new files leaving the original PDF unmodified.


Notes
-----
The scrubbing feature (-s) should not be trusted for any serious security
uses.  After using this experimental feature, please verify that it in fact
zero'd all of the objects that were of concern (those objects that were to be
zero'd).  Currently this feature will likely not render a working pdf.

This tool relies on the application reading the pdfresurrect extracted versions
to treat the last xref table as the most recent in the document.  This should
typically be the case.

The verbose output, which tries to deduce the PDF object type (e.g. stream,
page), is not always accurate, and the object counts might not be 100%
accurate.  However, this should not prevent the extraction of the versions.
This output is merely to provide a hint for the user as to what might be
different between the documents.

Object counts might appear off in linearized PDF documents.  That is not truly
the case, the reason for this is that each version of the PDF consists of the
objects that compose the linear portion of the PDF plus all of the objects that
compose the version in question.  Suppose there is a linearized PDF with 59
objects in its linear portion, and suppose the PDF has a second version that
consists of 21 objects.  The total number of objects in "version 2"
would be 59 + 21 or 80 objects.


Building
--------
From the top-level directory of pdfresurrect run:
    ./configure
    make

To install/uninstall the resulting binary to a specific path
the '--prefix=' flag can be used:
    ./configure --prefix=/my/desired/path/

Debugging mode can be enabled when configuring by using the following option:
    ./configure --enable-debug

The resulting binary can be placed anywhere, however it can also be
installed/uninstalled to the configured path automatically.  If no path was
specified at configure time, the default is /usr/local/bin
To install/uninstall:
    make install 
         or
    make uninstall


Thanks
------
The rest of the 757/757Labs crew.
GNU (www.gnu.org).
All of the contributors: See AUTHORS file.


Contact / Project URL
---------------------
mattdavis9@gmail.com
https://github.com/enferex/pdfresurrect


ChangeLog
-- Version 0.23 --
* Changed license from GPLv3 to BSD 3-Clause.

* pdf.c: Fix potential infinite loop from unchecked fseek.

-- Version 0.22 --
* pdf.c: Guard the xref plaintext scan.

-- Version 0.21 --
* pdfresurrect.1, pdf.c, main.c: Make scrub mode (-s) experimental.

* pdf.c: Update PDF header sanity check.

-- Version 0.20 --
* pdf.c: Make buf_idx a size_t.
  Guard strtok access (thanks rootup).
	Add a null check to the page load routine.
	Add a .gitignore (thanks rwhitworth).
	Clean up a strlen calc (thanks Francois Marier).
	Increase the creator buffer to 64B and guard against overflow (thanks
  carter-yagemann)

-- Version 0.19 --
* pdf.c: Add bounds guards for reading creator data.

-- Version 0.18 --

2019-08-10 Matt Davis
* pdf.c: Add a safe memory allocation routine (zeros and sanity checks).
				 Free a memory leak.
         Thanks to rtfingc for reporting the related bugs this patch
				 addresses.

-- Version 0.17 --

2019-08-06 Matt Davis
* pdf.c: Fix a buffer corruption bug identified by:
         capcorpscert0188, LeeSunWoo, ParkInChul

-- Version 0.16 --

2019-07-25 Matt Davis
* pdf.c: Fix a buffer corruption bug identified by j0lamma.

* main.c: Fix grammar in an error message.

-- Version 0.15 --

2018-05-08 Matt Davis <mattdavis9@gmail.com>

* pdf.c: Sanity check various file operations (fread, fseek).

* configure.ac: Remove GNU defaults.
*               User can override CFLAGS during configure.

2018-05-07 Michal Ambroz

* Makefile.in: Create the binary destination directory if it does not exist.

-- Version 0.14 --

2017-11-24 Matt Davis <mattdavis9@gmail.com>

* main.c: Simplify the writing of older version.
*         Only emit a startxref block with an older xref
*         position, and nothing more.

-- Version 0.13 --

2017-11-24 Matt Davis <mattdavis9@gmail.com>

* pdf.c: Bug fix - guard xref access during scrubbing.

2013-11-17 Matt Davis <mattdavis9@gmail.com>
* Makefile.in: Add the man1 directory if it does not exit.

2013-11-7 Ryan Schmidt <github@ryandesign.com>
* Makefile.in: Honor configure argument:
*              --exec-prefix, --bindir, --datarootdir, --mandir

2013-10-25: David Binderman.
* pdf.c: Added a logical fix submitted by David B.

-- Version 0.12 --

2012-11-30 Matt Davis <mattdavis9@gmail.com>

* pdfresurrect.1: Less emphasis on scrubbing.
*                 Cleaned up some wording.

2012-11-29 Matt Davis <mattdavis9@gmail.com>

* pdf.c: Use a better EOF finding algorithm.
*        This should ensure that all EOFs are found, previously if an EOF was
*        on a 256 character boundary, it would not be detected.

2012-11-25 Matt Davis <mattdavis9@gmail.com>

* configure.ac: Disable all optimizations when compiling a debugging version

* pdf.c: Clean up a comment


-- Version 0.11 --

2012-05-28 Francois Marier <francois@debian.org>

* Makefile.in: Add LDFLAGS macro
*
*              Update install and uninstall targets using DESTDIR macro
*
*              Update CFLAGS macro to include CPPFLAGS

* configure.ac: Preserve existing CFLAGS

2012-05-28 Matt Davis <mattdavis9@gmail.com>

* pdf.c: Prevent invalid access and double frees when the xref table is
*        of a zero length (thanks valgrind, oh so sexxy).
*
* ChangeLog: Correct a line not prefixed with an asterisk.

2012-04-29 Matt Davis <mattdavis9@gmail.com>

* pdf.c: Initialize a variable
*
*        Add a macro for safe file access

2010-03-21 Matt Davis <mattdavis9@gmail.com>

* pdf.c: Clean up memory leak just introduced by the new creator-from-object
*        piece
*
*        Removed strdup and used strndup for portability
*
*        Allow creation tags to be picked up even if they are stored in objects
*        and not inline.
*
*        Initialize a pointer


-- Version 0.10 --

2010-03-20 Matt Davis <mattdavis9@gmail.com>

* main.c: Add additional copyright year

* pdf.c: Allow for creation data to be pulled from objects as well as inline
*        Removed strdup and used strncpy for portability purposes

* Makefile.in: Allow for EXTRA_CFLAGS to be specified from command line

* pdfresurrect.1: Updated AUTHORS section to give Francois more props
*	          Adjust date


-- Version 0.9 --

2009-11-11 Matt Davis <mattdavis9@gmail.com>

* pdf.c: Display the entire line for creator information
*        Avoid terminating the creator information past the array index


-- Version 0.8 --

2009-09-10 Matt Davis <mattdavis9@gmail.com>

* pdf.c, main.c: Avoid stalling if we cannot find a trailer for linearized
*                PDFs.  This is probably because the PDF is using xref
*                streams.  Thanks to Francois for mentioning a stall was
*                occurring, and then providing the suspect document that caused
*                such a case.


-- Version 0.7 --

2009-09-09 Matt Davis <mattdavis9@gmail.com>

* configure.ac: Added function check for strdup and strtol

* pdf.{c,h}: Support to extract PDF creator information (not XML format)

* pdf.c: Count objects from linearized PDFs appropriately


2009-05-24 Francois Marier <francois@debian.org>

* pdfresurrect.1: Added Francois Marier's update to the man page
*                 escaping a hyphen


-- Version 0.6 --

2009-05-23 Matt Davis <mattdavis9@gmail.com> and

* pdfresurrect.1: Added Francois Marier's man page.
*                 Added NOTES and COPYRIGHT to man page.

* Makefile.in: Added .PHONY target for uninstall
*              Removed backup target
*              install and uninstall targets also add/remove the man page

* configure.ac: Added default prefix as suggested by Francois Marier


2009-05-22 Francois Marier <francois@debian.org>

* pdfresurrect.1: Added Francois Marier's man page.

* Makefile.in: Added Francois Marier's changes to add
*              install, distclean, and .PHONY targets


-- Version 0.5 --

2009-05-19 Matt Davis <mattdavis9@gmail.com>

* pdf.h: Added notes that PDFs of version 1.5+ can have compressed stream xref

* pdf.c: Warn on encoded xref streams
*

pdf.h
/******************************************************************************
 * pdf.h
 *
 * pdfresurrect - PDF history extraction tool
 *
 * Copyright (C) 2008, 2009, 2012 Matt Davis (enferex).
 *
 * Special thanks to all of the contributors:  See AUTHORS.
 *
 * Special thanks to 757labs (757 crew), they are a great group
 * of people to hack on projects and brainstorm with.
 *
 * pdf.h is part of pdfresurrect.
 * pdfresurrect is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * pdfresurrect is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with pdfresurrect.  If not, see <http://www.gnu.org/licenses/>.
 *****************************************************************************/

#ifndef PDF_H_INCLUDE
#define PDF_H_INCLUDE

#include <stdio.h>


/* Bit-maskable flags */
typedef unsigned short pdf_flag_t;
#define PDF_FLAG_NONE         0
#define PDF_FLAG_QUIET        1
#define PDF_FLAG_DISP_CREATOR 2


/* Generic key/value structure */
#define KV_MAX_KEY_LENGTH   32
#define KV_MAX_VALUE_LENGTH 128
typedef struct _kv_t
{
    char key[KV_MAX_KEY_LENGTH];
    char value[KV_MAX_VALUE_LENGTH];
} kv_t;


/* Information about who/what created the PDF
 * From 1.7 Spec for non-metadata entries
 */
typedef kv_t pdf_creator_t;


typedef struct _xref_entry
{
    int obj_id;
    long offset;
    int gen_num;
    char f_or_n;
} xref_entry_t;


typedef struct _xref_t
{
    long start;
    long end;

    /* Array of metadata about the pdf */
    pdf_creator_t *creator;
    int n_creator_entries;

    int n_entries;
    xref_entry_t *entries;


    /* PDF 1.5 or greater: xref can be encoded as a stream */
    int is_stream;

    /* If the PDF is linear multiple xrefs make up one single version */
    int is_linear;

    /* Version of the document this xref belongs */
    int version;
} xref_t;


typedef struct _pdf_t
{
    char  *name;
    short  pdf_major_version;
    short  pdf_minor_version;

    int     n_xrefs;
    xref_t *xrefs;

    /* PDF 1.5 or greater: xref can be encoded as a stream */
    int has_xref_streams;
} pdf_t;


extern pdf_t *pdf_new(const char *name);
extern void pdf_delete(pdf_t *pdf);

extern int pdf_is_pdf(FILE *fp);
extern void pdf_get_version(FILE *fp, pdf_t *pdf);

extern int pdf_load_xrefs(FILE *fp, pdf_t *pdf);

extern char pdf_get_object_status(
    const pdf_t *pdf,
    int          xref_idx,
    int          entry_idx);

extern void pdf_zero_object(
    FILE        *fp,
    const pdf_t *pdf,
    int          xref_idx,
    int          entry_idx);

extern void pdf_summarize(
    FILE        *fp,
    const pdf_t *pdf,
    const char  *name,
    pdf_flag_t   flags);

/* Returns '1' if we successfully display data (means its probably not xml) */
extern int pdf_display_creator(const pdf_t *pdf, int xref_idx);


#endif /* PDF_H_INCLUDE */


main.h
/******************************************************************************
 * main.h
 *
 * pdfresurrect - PDF history extraction tool
 * https://github.com/enferex/pdfresurrect
 *
 * See https://github.com/enferex/pdfresurrect/blob/master/LICENSE for license
 * information.
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Special thanks to all of the contributors:  See AUTHORS.
 * Special thanks to 757labs (757 crew), they are a great group
 * of people to hack on projects and brainstorm with.
 *****************************************************************************/

#ifndef MAIN_H_INCLUDE
#define MAIN_H_INCLUDE

#include <stdio.h>


#define EXEC_NAME "pdfresurrect"
#define VER_MAJOR "0"
#define VER_MINOR "24b"
#define VER       VER_MAJOR"."VER_MINOR


#define TAG "[pdfresurrect]"
#define ERR(...) {fprintf(stderr, TAG" -- Error -- " __VA_ARGS__);}

/* Returns a zero'd buffer of 'size' bytes or exits in failure. */
extern void *safe_calloc(size_t bytes);

#endif /* MAIN_H_INCLUDE */


main.c
/******************************************************************************
 * main.c
 *
 * pdfresurrect - PDF history extraction tool
 *
 * Copyright (C) 2008-2010, 2012, 2017, 2019-21 Matt Davis (enferex).
 *
 * Special thanks to all of the contributors:  See AUTHORS.
 *
 * Special thanks to 757labs (757 crew), they are a great group
 * of people to hack on projects and brainstorm with.
 *
 * main.c is part of pdfresurrect.
 * pdfresurrect is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * pdfresurrect is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with pdfresurrect.  If not, see <http://www.gnu.org/licenses/>.
 *****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <sys/stat.h>
#include <sys/types.h>
#include "main.h"
#include "pdf.h"


static void usage(void)
{
    printf(EXEC_NAME " Copyright (C) 2008-2010, 2012, 2013, 2017, 2019-21 "
           "Matt Davis (enferex)\n"
           "Special thanks to all contributors and the 757 crew.\n"
           "See the AUTHORS file for a list of other contributors.\n"
           "This program comes with ABSOLUTELY NO WARRANTY\n"
           "This is free software, and you are welcome to redistribute it\n"
           "under certain conditions.  For details see the file 'LICENSE'\n"
           "that came with this software or visit:\n"
           "<http://www.gnu.org/licenses/gpl-3.0.txt>\n\n");

    printf("-- " EXEC_NAME " v" VER" --\n"
           "Usage: ./" EXEC_NAME " <file.pdf> [-i] [-w] [-q]\n"
           "\t -i Display PDF creator information\n"
           "\t -w Write the PDF versions and summary to disk\n"
           "\t -q Display only the number of versions contained in the PDF\n");
// Experimental feature:
//           "\t -s Scrub the previous history data from the specified PDF\n");

    exit(0);
}


static void write_version(
    FILE       *fp,
    const char *fname,
    const char *dirname,
    xref_t     *xref)
{
    long  start;
    char *c, *new_fname, data;
    FILE *new_fp;

    start = ftell(fp);

    /* Create file */
    if ((c = strstr(fname, ".pdf")))
      *c = '\0';
    new_fname = safe_calloc(strlen(fname) + strlen(dirname) + 32);
    snprintf(new_fname, strlen(fname) + strlen(dirname) + 32,
             "%s/%s-version-%d.pdf", dirname, fname, xref->version);

    if (!(new_fp = fopen(new_fname, "w")))
    {
        ERR("Could not create file '%s'\n", new_fname);
        fseek(fp, start, SEEK_SET);
        free(new_fname);
        return;
    }

    /* Copy original PDF */
    fseek(fp, 0, SEEK_SET);
    while (fread(&data, 1, 1, fp))
      fwrite(&data, 1, 1, new_fp);

    /* Emit an older startxref, referring to an older version. */
    fprintf(new_fp, "\r\nstartxref\r\n%ld\r\n%%%%EOF", xref->start);

    /* Clean */
    fclose(new_fp);
    free(new_fname);
    fseek(fp, start, SEEK_SET);
}


#ifdef PDFRESURRECT_EXPERIMENTAL
static void scrub_document(FILE *fp, const pdf_t *pdf)
{
    FILE *new_fp;
    int   ch, i, j, last_version ;
    char *new_name, *c;
    const char *suffix = "-scrubbed.pdf";

    printf("The scrub feature (-s) is experimental and likely not to work as "
           "expected.\n");

    /* Create a new name */
    if (!(new_name = malloc(strlen(pdf->name) + strlen(suffix) + 1)))
    {
        ERR("Insufficient memory to create scrubbed file name\n");
        return;
    }

    strcpy(new_name, pdf->name);
    if ((c = strrchr(new_name, '.')))
      *c = '\0';
    strcat(new_name, suffix);

    if ((new_fp = fopen(new_name, "r")))
    {
        ERR("File name already exists for saving scrubbed document\n");
        free(new_name);
        fclose(new_fp);
        return;
    }

    if (!(new_fp = fopen(new_name, "w+")))
    {
        ERR("Could not create file for saving scrubbed document\n");
        free(new_name);
        fclose(new_fp);
        return;
    }

    /* Copy */
    fseek(fp, SEEK_SET, 0);
    while ((ch = fgetc(fp)) != EOF)
      fputc(ch, new_fp);

    /* Find last version (don't zero these baddies) */
    last_version = 0;
    for (i=0; i<pdf->n_xrefs; i++)
      if (pdf->xrefs[i].version)
        last_version = pdf->xrefs[i].version;

    /* Zero mod objects from all but the most recent version
     * Zero del objects from all versions
     */
    fseek(new_fp, 0, SEEK_SET);
    for (i=0; i<pdf->n_xrefs; i++)
    {
        for (j=0; j<pdf->xrefs[i].n_entries; j++)
          if (!pdf->xrefs[i].entries[j].obj_id)
            continue;
          else
          {
              switch (pdf_get_object_status(pdf, i, j))
              {
                  case 'M':
                      if (pdf->xrefs[i].version != last_version)
                        pdf_zero_object(new_fp, pdf, i, j);
                      break;

                  case 'D':
                      pdf_zero_object(new_fp, pdf, i, j);
                      break;

                  default:
                      break;
              }
          }
    }

    /* Clean */
    free(new_name);
    fclose(new_fp);
}
#endif // PDFRESURRECT_EXPERIMENTAL


static void display_creator(FILE *fp, const pdf_t *pdf)
{
    int i;

    printf("PDF Version: %d.%d\n",
           pdf->pdf_major_version, pdf->pdf_minor_version);

    for (i=0; i<pdf->n_xrefs; ++i)
    {
        if (!pdf->xrefs[i].version)
          continue;

        if (pdf_display_creator(pdf, i))
          printf("\n");
    }
}


static pdf_t *init_pdf(FILE *fp, const char *name)
{
    pdf_t *pdf;

    pdf = pdf_new(name);
    pdf_get_version(fp, pdf);
    if (pdf_load_xrefs(fp, pdf) == -1) {
      pdf_delete(pdf);
      return NULL;
    }

    return pdf;
}


void *safe_calloc(size_t size) {
  void *addr;

  if (!size)
  {
    ERR("Invalid allocation size.\n");
    exit(EXIT_FAILURE);
  }
  if (!(addr = calloc(1, size)))
  {
      ERR("Failed to allocate requested number of bytes, out of memory?\n");
      exit(EXIT_FAILURE);
  }
  return addr;
}


int main(int argc, char **argv)
{
    int         i, n_valid, do_write, do_scrub;
    char       *c, *dname, *name;
    DIR        *dir;
    FILE       *fp;
    pdf_t      *pdf;
    pdf_flag_t  flags;

    if (argc < 2)
      usage();

    /* Args */
    do_write = do_scrub = flags = 0;
    name = NULL;
    for (i=1; i<argc; i++)
    {
        if (strncmp(argv[i], "-w", 2) == 0)
          do_write = 1;
        else if (strncmp(argv[i], "-i", 2) == 0)
          flags |= PDF_FLAG_DISP_CREATOR;
        else if (strncmp(argv[i], "-q", 2) == 0)
          flags |= PDF_FLAG_QUIET;
#ifdef PDFRESURRECT_EXPERIMENTAL
        else if (strncmp(argv[i], "-s", 2) == 0)
          do_scrub = 1;
#endif
        else if (argv[i][0] != '-')
          name = argv[i];
        else if (argv[i][0] == '-')
          usage();
    }

    if (!name)
      usage();

    if (!(fp = fopen(name, "r")))
    {
        ERR("Could not open file '%s'\n", argv[1]);
        return -1;
    }
    else if (!pdf_is_pdf(fp))
    {
        ERR("'%s' specified is not a valid PDF\n", name);
        fclose(fp);
        return -1;
    }

    /* Load PDF */
    if (!(pdf = init_pdf(fp, name)))
    {
        fclose(fp);
        return -1;
    }

    /* Count valid xrefs */
    for (i=0, n_valid=0; i<pdf->n_xrefs; i++)
      if (pdf->xrefs[i].version)
        ++n_valid;

    /* Bail if we only have 1 valid */
    if (n_valid < 2)
    {
        if (!(flags & (PDF_FLAG_QUIET | PDF_FLAG_DISP_CREATOR)))
          printf("%s: There is only one version of this PDF\n", pdf->name);

        if (do_write)
        {
            fclose(fp);
            pdf_delete(pdf);
            return 0;
        }
    }

    dname = NULL;
    if (do_write)
    {
        /* Create directory to place the various versions in */
        if ((c = strrchr(name, '/')))
          name = c + 1;

        if ((c = strrchr(name, '.')))
          *c = '\0';

        dname = safe_calloc(strlen(name) + 16);
        sprintf(dname, "%s-versions", name);
        if (!(dir = opendir(dname)))
          mkdir(dname, S_IRWXU);
        else
        {
            ERR("This directory already exists, PDF version extraction will "
                "not occur.\n");
            fclose(fp);
            closedir(dir);
            free(dname);
            pdf_delete(pdf);
            return -1;
        }

        /* Write the pdf as a previous version */
        for (i=0; i<pdf->n_xrefs; i++)
          if (pdf->xrefs[i].version)
            write_version(fp, name, dname, &pdf->xrefs[i]);
    }

    /* Generate a per-object summary */
    pdf_summarize(fp, pdf, dname, flags);

#ifdef PDFRESURRECT_EXPERIMENTAL
    /* Have we been summoned to scrub history from this PDF */
    if (do_scrub)
      scrub_document(fp, pdf);
#endif

    /* Display extra information */
    if (flags & PDF_FLAG_DISP_CREATOR)
      display_creator(fp, pdf);

    fclose(fp);
    free(dname);
    pdf_delete(pdf);

    return 0;
}
